pkg/app/api/grpcapi/piped_api.go

```diff
@@ -1105,6 +1105,54 @@ func (a *PipedAPI) CreateDeploymentChain(ctx context.Context, req *pipedservice.
 	return &pipedservice.CreateDeploymentChainResponse{}, nil
 }
 
+// InChainDeploymentPlannable hecks the completion and status of the previous block in the deployment chain.
+// An in chain deployment is treated as plannable in case:
+// - It's the first deployment of its deployment chain.
+// - All deployments of its previous block in chain are at DEPLOYMENT_SUCCESS state.
+func (a *PipedAPI) InChainDeploymentPlannable(ctx context.Context, req *pipedservice.InChainDeploymentPlannableRequest) (*pipedservice.InChainDeploymentPlannableResponse, error) {
+	_, pipedID, _, err := rpcauth.ExtractPipedToken(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if err := a.validateDeploymentBelongsToPiped(ctx, req.Deployment.Id, pipedID); err != nil {
+		return nil, err
+	}
+
+	dc, err := a.deploymentChainStore.GetDeploymentChain(ctx, req.Deployment.DeploymentChainId)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, "unable to find the deployment chain which this deployment belongs to")
+	}
+
+	// Deployment of blocks[0] in the chain means it's the first deployment of the chain;
+	// hence it should be processed without any lock.
+	if req.Deployment.DeploymentChainBlockIndex == 0 {
+		return &pipedservice.InChainDeploymentPlannableResponse{
+			Plannable: true,
+		}, nil
+	}
+
+	if req.Deployment.DeploymentChainBlockIndex >= int32(len(dc.Blocks)) {
+		return nil, status.Error(codes.InvalidArgument, "invalid deployment with chain block index provided")
+	}
+
+	prevBlock := dc.Blocks[req.Deployment.DeploymentChainBlockIndex-1]
+	plannable := true
+	for _, node := range prevBlock.Nodes {
+		dp, err := a.deploymentStore.GetDeployment(ctx, node.DeploymentRef.DeploymentId)
```

Comment:
I think we need the deployment status in the chain model for rendering on the web so how about having that `status` field in the node. Then we can use it directly instead of retrieving from DB like this.

Reply:
In that case, when piped submit deployment status (planed, running, etc) we also need to update its status in deployment chain model too, that's quite annoying. How do you think about adding just `IsCompleted` to the deploymentRef in the chain model?

Reply:
In the second thought, `status` is required for showing in the UI, so I think I will go with your suggestion 👍

Reply:
@nghialv Can I add a TODO note here and address it later since that change requires updates in other RPCs unrelated to this PR.

Reply:
Addressed by https://github.com/pipe-cd/pipe/pull/2832/commits/e883d32181fb593cb6cf679d6394e219b8215963